### DISCUSSION TITLE
Adding Support for Hebrew

### DISCUSSION THREAD

**Discussion by eliorc** [explosion/spaCy#8234] [Language Support]
Hi, I've already read the [Adding models for new languages master thread](https://github.com/explosion/spaCy/discussions/3056) and I couldn't find a straightforward guide for adding a new language.

To be honest, adding Hebrew to spaCy is an initiative I've tried to take before, but that time (few years ago) the language resources were very poor.
Lately there have been advancement (models for PoS, language models etc.) which are created by the community and also some initiatives that are supported (also econimically) by the goverment which makes adding Hebrew to spaCy much more feasible.

My question is, where can I find a documented step by step guide for adding a new language from scratch?

(I am aware of the [he section](https://github.com/explosion/spaCy/tree/master/spacy/lang/he) in the languages)

**Comment by polm** [comment]
Sorry, we don't have a better guide than that thread. 

What data do you have?

Since the basic tokenizer support for Hebrew is already in spaCy, if you have CoNLL format data with tagger and parser annotations, you should be able to use `spacy convert` and `spacy train` to train a model. That would be a good first step.

**Reply by eliorc** [reply]
We do have conllu format data ([link](https://github.com/UniversalDependencies/UD_Hebrew-HTB)) - what do you mean by tagger and parser? Are we talking about dependency parsing and PoS tagging? Sorry if this a dumb question, I'm not that familiar with the internals

My quest for step by step guide, because there are some really motivated people to make this happen, so if we can figure out what are the steps necessary I believe it's posssible to allocate resources to make it happen

**Reply by polm** [reply]
> what do you mean by tagger and parser? Are we talking about dependency parsing and PoS tagging? 

Yes. You can read about the default spaCy pipeline components [here](https://spacy.io/usage/processing-pipelines#built-in). 

Since you have CoNLL data, I recommend you take that data and use `spacy convert` to convert it to the spaCy binary format, and then use the [quickstart](https://spacy.io/usage/training#quickstart) to try training the tagger and parser. It'll probably be easier to do one at a time first to make sure everything is working. 

For a full model we'll also want NER data, which is generally not included in the UD datasets, and doesn't seem to be included in yours. So that's something you'll need to look for.

You'll also want word vectors. If there's a pretrained set somewhere with a good license that can be re-used, but if not something can be trained from Wikipedia or Common Crawl or another source. It looks like [Fasttext has vectors](https://fasttext.cc/docs/en/crawl-vectors.html) so you can just use them if there's nothing wrong with them.

Besides that you'll want a lemmatizer, tokenizer exceptions, and a tag map or other way to convert from fine-grained tags to coarse-grained tags. The best advice I can give for these is to look at what other languages do in spaCy and copy as appropriate.

**Reply by eliorc** [reply]
> Since you have CoNLL data, I recommend you take that data and use spacy convert to convert it to the spaCy binary format, and then use the quickstart to try training the tagger and parser. It'll probably be easier to do one at a time first to make sure everything is working.

I'll go ahead and try that, thanks

> For a full model we'll also want NER data, which is generally not included in the UD datasets, and doesn't seem to be included in yours. So that's something you'll need to look for.

There is a very fresh NER model implementation for Hebrew, called [NEMO](https://github.com/OnlpLab/NEMO) - is it better to use the dataset it was trained on, or is it possible to incorporate the model directly?

> You'll also want word vectors. If there's a pretrained set somewhere with a good license that can be re-used, but if not something can be trained from Wikipedia or Common Crawl or another source. It looks like Fasttext has vectors so you can just use them if there's nothing wrong with them.

Easy, we have tons of these :D 

> Besides that you'll want a lemmatizer, tokenizer exceptions, and a tag map or other way to convert from fine-grained tags to coarse-grained tags. The best advice I can give for these is to look at what other languages do in spaCy and copy as appropriate.

I'll have to further explore this and even if there are gaps, I'm sure we'll be able to close it

**Reply by polm** [reply]
> There is a very fresh NER model implementation for Hebrew, called NEMO - is it better to use the dataset it was trained on, or is it possible to incorporate the model directly?

If you just want to use it, it's possible to wrap existing models for use in spaCy. For pretrained models we distribute, we only use our internal architecture, so you'll have to retrain it on the data.

**Reply by eliorc** [reply]
About the second point - does spaCy necessarily expects data to train NER, or is it possible to supply an already trained model?

**Reply by polm** [reply]
If you just want to use a model with spaCy, it's possible to wrap a model as a pipeline component and set entities. We don't have an explicit guide to this but if you look at the [custom components docs](https://spacy.io/usage/processing-pipelines#custom-components) I think it's pretty straightforward.

If you wrap a model that way you can host it somewhere and add it to the Universe of third-party packages.

For the official pretrained models we distribute, we need the training data for various reasons, such as retraining models when we update spaCy. 

Also, I forgot to mention this before, but we don't generally release official models using non-commercial datasets, which it looks like your CoNLL data is. So that could be an issue.

**Reply by eliorc** [reply]
> but we don't generally release official models using non-commercial datasets

Do you mean you require commercial datasets?

**Reply by polm** [reply]
The CoNLL dataset you linked to has a BY-NC-SA license, which means that it cannot be used in "commercial" contexts. spaCy itself is under the MIT license and can be used for any purpose, and many of our users use it in professional contexts. For that reason we prefer datasets without a "non-commercial use" restriction.

**Reply by eliorc** [reply]
Just updating, I'm doing my best efforts to get my hands on a commercial CoNLL dataset with as much features as possible as well as NER data.

**Reply by polm** [reply]
I think we are on the same page, but to be perfectly clear, we do not require a "commercial dataset", just a dataset that is OK for commercial use. 

I would use "commercial dataset" to describe a dataset that is only available for a fee. A dataset does not have to be commercial in that sense to be OK for commercial use.

**Reply by eliorc** [reply]
Just to make sure, data under the [Creative Commons Attribution 4.0 International Public License](https://creativecommons.org/licenses/by/4.0/) is fine for spaCy purposes?

**Reply by polm** [reply]
Yes, CC BY should be fine.

**Comment by thewh1teagle** [comment]
@eliorc 
Have you made any progress with it?

**Reply by eliorc** [reply]
Nope, I don't have enough time to put into this, so I've given up

**Reply by thewh1teagle** [reply]
I just started to learn AI with python (I already have a lot of experience in programming)
And I'm very interested in that project and any project related to NLP with hebrew.
I understand that if I want to make a project for that I must create word embedding for hebrew, And I think that's what you wanted to do?
Do you mind help me with that?

**Reply by eliorc** [reply]
Sorry, as I said I don't have the availability for that
